cirq-core/cirq/devices/device.py

```diff
@@ -253,3 +263,108 @@ def _from_json_dict_(cls, qubits, nx_graph, **kwargs):
         if nx_graph != '':
             graph_obj = nx.readwrite.json_graph.node_link_graph(nx_graph)
         return cls(qubits, graph_obj)
+
+
+@value.value_equality
+class GridDeviceMetadata(DeviceMetadata):
+    """Hardware metadata for homogenous 2d symmetric grid devices."""
+
+    def __init__(
+        self,
+        qubit_pairs: Iterable[Tuple['cirq.Qid', 'cirq.Qid']],
+        supported_gates: 'cirq.Gateset',
+        gate_durations: Optional[Dict['cirq.Gateset', 'cirq.Duration']] = None,
+    ):
+        """Create a GridDeviceMetadata object.
+
+        Create a GridDevice which has a well defined set of couplable
+        qubit pairs that have the same two qubit gates available in
+        both coupling directions.
+
+        Args:
+            qubit_pairs: Iterable of pairs of `cirq.Qid`s representing
+                bi-directional couplings.
+            supported_gates: `cirq.Gateset` indicating gates supported
+                everywhere on the device.
+            gate_durations: Optional dictionary of `cirq.Gateset`
+                instances mapping to `cirq.Duration` instances for
+                gate timing metadata information. If provided,
+                must match all entries in supported_gates.
+
+        Raises:
+            ValueError: if the union of gateset keys in gate_durations,
+                do not represent an identical gateset to supported_gates.
+        """
+        qubit_pairs = list(qubit_pairs)
+        flat_pairs = [q for pair in qubit_pairs for q in pair]
+        # Keep lexigraphically smaller tuples for undirected edges.
+        sorted_pairs = sorted(qubit_pairs)
+        qubit_set = set()
+        for a, b in sorted_pairs:
+            if (b, a) not in qubit_set:
+                qubit_set.add((a, b))
```

Comment:
Alternatively, `qubit_set = set(map(sorted, qubit_pairs))`

Also, the naming here is strange - `qubit_set` for the set of pairs, and `flat_pairs` for the set of qubits? Seems like it should be reversed.

Reply:
This doesn't work. Sorted converts the tuples to a list and then it's no longer hashable.